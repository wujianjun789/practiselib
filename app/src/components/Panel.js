/**
 * create by Azrael on 2017/02/21
 */
import React, {Component} from 'react';
import PropTypes from 'prop-types';
/**
 * Panel component
 * @param {String}      className   'panel theme,default:　panel-primary'
 * @param {String}      title       'panel title, default: '　'
 * @param {Any}         body        'panel body'
 * @param {Any}         footer      'panel footer,default:hidden'
 * @param {String}      text        'no body tips,default: '无相关数据!''
 * @param {Boolean}     closeBtn    'panel close,true: display, false: hidden, default: false'
 * @param {Func}        closeClick  'panel close button handler'
 * 
 */
const Panel = (props) => {
    let {className='', closeClick=null, title='', text, body, footer, closeBtn, children} = props;
    body = body ? body : <div className="row pull-center">{text ? text : '无相关数据'}</div>;
    footer = footer ? footer : null,
    closeBtn =  closeBtn ? true : false;
    return (
        <div className={`modal ${className}`} role="dialog">
            <div className="modal-dialog" role="document">
                <div className="modal-content">
                    <div className="modal-header">
                        {closeBtn ? <button type="button" className="close" data-dismiss="modal" aria-label="Close" onClick={closeClick}><span aria-hidden="true">&times;</span></button> : null}
                        <h4 className="modal-title">{title}</h4>
                    </div>
                    <div className="modal-body">
                        {children ? children : body }
                    </div>
                    { footer }
                </div>
            </div>
        </div>
    )
}

Panel.propTypes = {
    className: PropTypes.string,
    title: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.node
    ]),
    body: PropTypes.any,
    footer: PropTypes.any,
    text: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.node
    ]),
    closeBtn: PropTypes.bool,
    closeClick: PropTypes.func
}

export default Panel;