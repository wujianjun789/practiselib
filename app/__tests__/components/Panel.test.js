import React from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import Panel from '../../src/components/Panel';

describe('<Panel />', () => {
    it('render with no props passed', () => {
        const panel = shallow(<Panel />);

        const container = panel.find('.modal-content');
        expect(container.length).toBe(1);

        const tit = panel.find('.modal-title');
        expect(tit.length).toBe(1);
        expect(tit.text()).toBe('');

        expect(panel.find('button.close').length).toBe(0);

        const body = panel.find('.modal-body');
        expect(body.length).toBe(1);
        const bodyChild = body.find('.row.pull-center');
        expect(bodyChild.length).toBe(1);
        expect(bodyChild.text()).toBe('无相关数据');
    });

    it('render with className="modal-default", title=box, text="body text", closeBtn=true', () => {
        const panel = shallow(<Panel className="modal-default" title="box" text="body text" closeBtn/>);

        const container = panel.find('.modal-content');
        expect(container.hasClass('modal-default')).toBeTruthy();

        const tit = panel.find('.modal-title');
        expect(tit.text()).toBe('box');

        expect(panel.find('button.close').length).toBe(1);

        const body = panel.find('.modal-body');
        const bodyChild = body.find('.row.pull-center');
        expect(bodyChild.text()).toBe('body text');
    });

    it('render with body=<div className="test">test</div>, footer=<div className="footer">footer</div>', () => {
        let body = <div className="test">test</div>;
        let footer = <div className="footer">footer</div>;
        const panel = shallow(<Panel body={body} footer={footer}/>);
        
        const bod = panel.find('.test');
        expect(bod.length).toBe(1);
        expect(bod.text()).toBe('test');

        const foot = panel.find('.footer');
        expect(foot.length).toBe(1);
        expect(foot.text()).toBe('footer');
    });

    it('render with children=<div className="test">test</div>', () => {
        const panel = mount(<Panel><div className="test">test</div></Panel>);

        const body = panel.find('.test');
        expect(body.length).toBe(1);
        expect(body.text()).toBe('test');
    });

    it('closeBtn click', () => {
        const click = jest.fn();
        const panel = shallow(<Panel closeClick={click} closeBtn />);
        let closeBtn = panel.find('button.close');
        closeBtn.simulate('click');
        expect(click).toHaveBeenCalledTimes(1);
    })

    it('snapshot with no props passed', () => {
        let cmp = renderer.create(<Panel />);
        let tree = cmp.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('snapshot with className="modal-default", title=box, text="body text", closeBtn=true', () => {
        const panel = renderer.create(<Panel className="modal-default" title="box" text="body text" closeBtn/>);
        let tree = panel.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('snapshot with body=<div className="test">test</div>, footer=<div className="footer">footer</div>', () => {
        let body = <div className="test">test</div>;
        let footer = <div className="footer">footer</div>;
        const panel = renderer.create(<Panel body={body} footer={footer}/>);
        let tree = panel.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('snapshot with children=<div className="test">test</div>', () => {
        let click = jest.fn();
        const panel = renderer.create(<Panel closeClick={click} closeBtn><div className="test">test</div></Panel>);
        let tree = panel.toJSON();
        expect(tree).toMatchSnapshot();
    });
})