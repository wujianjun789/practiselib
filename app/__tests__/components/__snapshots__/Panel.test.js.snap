exports[`<Panel /> snapshot with body=<div className="test">test</div>, footer=<div className="footer">footer</div> 1`] = `
<div
  className="modal "
  role="dialog">
  <div
    className="modal-dialog"
    role="document">
    <div
      className="modal-content">
      <div
        className="modal-header">
        <h4
          className="modal-title">
          
        </h4>
      </div>
      <div
        className="modal-body">
        <div
          className="test">
          test
        </div>
      </div>
      <div
        className="modal-footer">
        <div
          className="footer">
          footer
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Panel /> snapshot with children=<div className="test">test</div> 1`] = `
<div
  className="modal "
  role="dialog">
  <div
    className="modal-dialog"
    role="document">
    <div
      className="modal-content">
      <div
        className="modal-header">
        <button
          aria-label="Close"
          className="close"
          data-dismiss="modal"
          onClick={[Function]}
          type="button">
          <span
            aria-hidden="true">
            ×
          </span>
        </button>
        <h4
          className="modal-title">
          
        </h4>
      </div>
      <div
        className="modal-body">
        <div
          className="test">
          test
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Panel /> snapshot with className="modal-default", title=box, text="body text", closeBtn=true 1`] = `
<div
  className="modal modal-default"
  role="dialog">
  <div
    className="modal-dialog"
    role="document">
    <div
      className="modal-content">
      <div
        className="modal-header">
        <button
          aria-label="Close"
          className="close"
          data-dismiss="modal"
          onClick={null}
          type="button">
          <span
            aria-hidden="true">
            ×
          </span>
        </button>
        <h4
          className="modal-title">
          box
        </h4>
      </div>
      <div
        className="modal-body">
        <div
          className="row pull-center">
          body text
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Panel /> snapshot with no props passed 1`] = `
<div
  className="modal "
  role="dialog">
  <div
    className="modal-dialog"
    role="document">
    <div
      className="modal-content">
      <div
        className="modal-header">
        <h4
          className="modal-title">
          
        </h4>
      </div>
      <div
        className="modal-body">
        <div
          className="row pull-center">
          无相关数据
        </div>
      </div>
    </div>
  </div>
</div>
`;
